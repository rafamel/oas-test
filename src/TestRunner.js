import deep from 'lodash.clonedeep';
// sync alternative: json-ref-lite
import $RefParser from 'json-schema-ref-parser';
import SwaggerParser from 'swagger-parser';
import chalk from 'chalk';
import collect from './collect';

export const STORE_SYMBOL = Symbol('store');

export default class TestRunner {
  constructor() {
    this._first = false;
    this[STORE_SYMBOL] = {
      app: null,
      url: null,
      settings: {
        assert: (bool) => expect(bool).toBe(true),
        describe: describe,
        test: test,
        logger: console,
        logall: false
      },
      hooks: {
        data: [],
        request: [],
        response: []
      },
      context: {},
      oas: Promise.resolve(null),
      collection: Promise.resolve(null),
      assertlog: (condition, description) => {
        const { settings } = this[STORE_SYMBOL];
        if (
          settings.logger &&
          !condition &&
          (settings.logall || !this._first)
        ) {
          this._first = true;
          settings.logger.log(chalk.red(description));
        }
        settings.assert(condition);
      }
    };
  }
  get context() {
    return this[STORE_SYMBOL].context;
  }
  load({ app, url, oas } = {}) {
    this[STORE_SYMBOL].app = app || null;
    this[STORE_SYMBOL].url = url || null;

    if (oas) {
      // Dereference and validate oas
      const promise = Promise.resolve(oas)
        .then((oas) => $RefParser.dereference(deep(oas)))
        .then((oas) => SwaggerParser.validate(oas))
        .catch((e) => {
          throw e;
        });
      this[STORE_SYMBOL].oas = promise;
      this[STORE_SYMBOL].collection = promise.then(collect);
    } else {
      this[STORE_SYMBOL].oas = Promise.resolve(null);
      this[STORE_SYMBOL].collection = Promise.resolve(null);
    }

    return this;
  }
  settings(obj = {}) {
    Object.assign(this[STORE_SYMBOL].settings, obj);
    return this;
  }
  hook(type, cb) {
    const hookTypes = ['data', 'request', 'response'];
    if (!hookTypes.includes(type)) {
      throw Error(`Hooks can only be of types ${hookTypes.join(', ')}.`);
    }
    this[STORE_SYMBOL].hooks[type].push(cb);
    return this;
  }
  clone() {
    const testRunner = new TestRunner();
    testRunner[STORE_SYMBOL] = deep(this[STORE_SYMBOL]);
    return testRunner;
  }
  self(cb) {
    // eslint-disable-next-line standard/no-callback-literal
    if (cb) cb(this);
    return this;
  }
  test(tests = {}) {
    const store = this[STORE_SYMBOL];

    if (!store.app && !store.url) {
      throw Error(
        "TestRunner.load() hasn't been called with a valid app, url, or both"
      );
    }

    return this;
  }
}
