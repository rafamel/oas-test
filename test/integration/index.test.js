import TestRunner from '../../src';
import app from '../setup/app';
import oas from '../setup/oas';
import jsf from 'json-schema-faker';
import basicTests from './basic-tests';
import appTests from './app-tests';
import appOasTests from './app-oas-tests';

// Set randomness generator for faker as constant
jsf.option({ random: () => 0.5 });

const rootTr = new TestRunner().load({ app, oas }).settings({
  logger: { log: () => {} },
  describe: (_, cb) => cb(),
  test: (_, cb) => cb()
});

basicTests(rootTr);
appTests(rootTr, app);
appOasTests(rootTr, app, oas);
