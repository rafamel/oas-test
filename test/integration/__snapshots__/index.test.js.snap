// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`w/ app & oas Callbacks receive all args 1`] = `
Object {
  "all": Object {
    "body": Object {
      "done": false,
      "name": "Lorem anim ipsum",
    },
    "cookies": Object {
      "aCookie": "Lorem anim ipsum",
    },
    "headers": Object {
      "some": "Lorem anim ipsum",
    },
    "params": Object {
      "id": 50000000,
    },
    "query": Object {
      "name": 50000000,
    },
  },
  "mimeTypes": Array [],
  "required": Object {
    "body": Object {
      "done": false,
      "name": "Lorem anim ipsum",
    },
    "cookies": Object {},
    "headers": Object {
      "some": "Lorem anim ipsum",
    },
    "params": Object {
      "id": 50000000,
    },
    "query": Object {},
  },
}
`;

exports[`w/ app & oas Callbacks receive all args 2`] = `
Object {
  "context": Object {},
  "endpoint": Object {
    "method": "put",
    "operation": null,
    "path": "/api/todo/{id}",
  },
  "id": "updateTodo",
  "iteration": 0,
  "name": "200",
  "oas": null,
}
`;

exports[`w/ app & oas Callbacks receive all args 3`] = `
Object {
  "data": [Function],
  "repeat": 1,
  "request": [Function],
  "response": [Function],
  "validate": null,
}
`;

exports[`w/ app & oas Callbacks receive all args 4`] = `
Object {
  "context": Object {},
  "endpoint": Object {
    "method": "put",
    "operation": null,
    "path": "/api/todo/{id}",
  },
  "id": "updateTodo",
  "iteration": 0,
  "name": "200",
  "oas": null,
}
`;

exports[`w/ app & oas Callbacks receive all args 5`] = `
Object {
  "data": [Function],
  "repeat": 1,
  "request": [Function],
  "response": [Function],
  "validate": null,
}
`;

exports[`w/ app & oas Callbacks receive all args 6`] = `
Object {
  "context": Object {},
  "endpoint": Object {
    "method": "put",
    "operation": null,
    "path": "/api/todo/{id}",
  },
  "id": "updateTodo",
  "iteration": 0,
  "name": "200",
  "oas": null,
}
`;

exports[`w/ app & oas Callbacks receive all args 7`] = `
Object {
  "data": [Function],
  "repeat": 1,
  "request": [Function],
  "response": [Function],
  "validate": null,
}
`;

exports[`w/ app & oas Generates fake data for header and body 1`] = `
Object {
  "all": Object {
    "body": Object {
      "done": false,
      "name": "Lorem anim ipsum",
    },
    "cookies": Object {
      "aCookie": "Lorem anim ipsum",
    },
    "headers": Object {
      "some": "Lorem anim ipsum",
    },
    "params": Object {
      "id": 50000000,
    },
    "query": Object {
      "name": 50000000,
    },
  },
  "mimeTypes": Array [],
  "required": Object {
    "body": Object {
      "done": false,
      "name": "Lorem anim ipsum",
    },
    "cookies": Object {},
    "headers": Object {
      "some": "Lorem anim ipsum",
    },
    "params": Object {
      "id": 50000000,
    },
    "query": Object {},
  },
}
`;

exports[`w/ app Callbacks receive all args 1`] = `
Object {
  "context": Object {},
  "endpoint": Object {
    "method": "get",
    "operation": null,
    "path": "/api/wappwooas",
  },
  "id": "get:/api/wappwooas",
  "iteration": 0,
  "name": "200",
  "oas": null,
}
`;

exports[`w/ app Callbacks receive all args 2`] = `
Object {
  "data": [Function],
  "repeat": 1,
  "request": [Function],
  "response": [Function],
  "validate": null,
}
`;

exports[`w/ app Callbacks receive all args 3`] = `
Object {
  "context": Object {},
  "endpoint": Object {
    "method": "get",
    "operation": null,
    "path": "/api/wappwooas",
  },
  "id": "get:/api/wappwooas",
  "iteration": 0,
  "name": "200",
  "oas": null,
}
`;

exports[`w/ app Callbacks receive all args 4`] = `
Object {
  "data": [Function],
  "repeat": 1,
  "request": [Function],
  "response": [Function],
  "validate": null,
}
`;

exports[`w/ app Callbacks receive all args 5`] = `
Object {
  "context": Object {},
  "endpoint": Object {
    "method": "get",
    "operation": null,
    "path": "/api/wappwooas",
  },
  "id": "get:/api/wappwooas",
  "iteration": 0,
  "name": "200",
  "oas": null,
}
`;

exports[`w/ app Callbacks receive all args 6`] = `
Object {
  "data": [Function],
  "repeat": 1,
  "request": [Function],
  "response": [Function],
  "validate": null,
}
`;
